README.md
[![NPM version](https://img.shields.io/npm/v/express-open-in-editor.svg)](https://www.npmjs.com/package/express-open-in-editor)
[![Build Status](https://travis-ci.org/lahmatiy/express-open-in-editor.svg?branch=master)](https://travis-ci.org/lahmatiy/express-open-in-editor)
[![Coverage Status](https://coveralls.io/repos/github/lahmatiy/express-open-in-editor/badge.svg?branch=master)](https://coveralls.io/github/lahmatiy/express-open-in-editor?branch=master)
[![Dependency Status](https://img.shields.io/david/lahmatiy/express-open-in-editor.svg)](https://david-dm.org/lahmatiy/express-open-in-editor)

Express middleware to open any file in an editor by request to defined route. Based on [open-in-editor](https://github.com/lahmatiy/open-in-editor).

## Install

```
npm install express-open-in-editor
```

## Usage

```js
var express = require('express');
var openInEditor = require('express-open-in-editor');

var app = express();

// There are few ways to setup:
// - to trigger middleware on *GET* request to `/open-in-editor`
app.get('/open-in-editor', openInEditor());

// - to trigger middleware on *any* request method to `/open-in-editor`
app.use('/open-in-editor', openInEditor());

// - to trigger middleware on *any* request method to *any* path
//   (not recommended unless server's single purpose is to open files in editor)
app.use(openInEditor());
```

After that you can use GET requests like `/open-in-editor?file=foo/bar.ext:2:5` to open `foo/bar.ext` in an editor at line 2 column 5.

By default `express-open-in-editor` uses `process.env.VISUAL` or `process.env.EDITOR` (with this priority) to define the command to open a file in an editor. It could be set globally or on script execution:

```
EDITOR=subl node app.js
```

Also you can set `process.env.OPEN_FILE` that has highest priority and understands shorthands (i.e. `subl` for `Sublime Text` or `atom` for `Atom Editor`).

For more details about setup see [open-in-editor](https://github.com/lahmatiy/open-in-editor) description.

### Using with webpack-dev-server

Although `webpack-dev-server` uses `express` to create a server, you have the same options to apply the middleware to it. The only difference is that you should define it in `setup` method (see [issue](https://github.com/webpack/webpack-dev-server/issues/285) for details).

```js
var server = new WebpackDevServer(webpack(config), {
  // ...
  setup: function(app) {
    app.use('/open-in-editor', openInEditor());
  }
});
```

## API

```
openInEditor([options]);
```

Options:

- `cwd`

  Type: `String`  
  Default: `''` (uses `process.cwd()`)

  Working directory used when resolving paths.

- Rest options are passing to [open-in-editor](https://github.com/lahmatiy/open-in-editor) as is.

## Related projects

- [open-in-editor](https://github.com/lahmatiy/open-in-editor) – package that do the main task of `express-open-in-editor`, i.e. opens file in editor.
- [babel-plugin-source-wrapper](https://github.com/restrry/babel-plugin-source-wrapper) – `Babel` plugin that instruments source code to associate objects with location they defined in code base.
- [Component Inspector](https://github.com/lahmatiy/component-inspector) – developer tool to inspect components that can open component creation source location in editor. Has integrations for `React`, `Backbone` and can be adopter for other frameworks.

## License

MIT


HISTORY.md
## 3.1.1 (November 10, 2017)

- Fixed absolute file path resolving on Windows (@bashmish, #7)

## 3.1.0 (October 1, 2017)

- Added `cwd` option
- Removed `parseurl` from dependencies

## 3.0.2 (September 21, 2017)

- Fixed crash on send error in response
- Fixed messages

## 3.0.1 (September 21, 2017)

- Fixed crash on configure error output

## 3.0.0 (September 21, 2017)

- Reworked to follow `express` best practices (#1)

## 2.0.0 (March 22, 2017)

- Bump `open-in-editor` to [2.0.0](https://github.com/lahmatiy/open-in-editor/releases/tag/v2.0.0)

## 1.2.1 (September 6, 2016)

- Fixed leading slash cut in absolute file path

## 1.2.0 (January 25, 2016)

- Bump `open-in-editor` to `1.2.0`

## 1.1.0 (November 8, 2015)

- Show warning for any failed request
- Bump `open-in-editor` version

## 1.0 (September 24, 2015)

- Initial release


test/index.js
var assert = require('assert');
var express = require('express');
var http = require('http');
var path = require('path');
var openInEditor = require('../index.js');
var PORT = process.env.PORT || 8999;
var OPEN_FILE = 'echo';

process.env.OPEN_FILE = OPEN_FILE;

function request(options, expectedStatusCode, expectedResponse, done) {
    return function(done) {
        http.request(Object.assign({
            hostname: 'localhost',
            port: PORT,
            method: 'GET'
        }, options), function(res) {
            var actualResponse = '';

            assert.equal(res.statusCode, expectedStatusCode);

            res.setEncoding('utf8');
            res.on('data', function(chunk) {
                actualResponse += chunk;
            });
            res.on('end', function() {
                if (typeof expectedResponse === 'string') {
                    assert.equal(actualResponse, expectedResponse);
                } else {
                    assert(expectedResponse.test(actualResponse), 'should match to ' + expectedResponse);
                }
                done();
            });
        }).end();
    };
}

describe('app.use(middleware)', function() {
    var server;
    var app;
    var tests = [
        [{ path: '/any?file=foo' }, 200, 'Open in editor: ' + path.join(process.cwd(), 'foo')],
        [{ path: '/any?file=foo', method: 'POST' }, 200, 'Open in editor: ' + path.join(process.cwd(), 'foo')],
        [{ path: '/any?filex=foo' }, 400, '[express-open-in-editor] Parameter missed: file']
    ];

    beforeEach(function(done) {
        app = express();
        app.use(openInEditor());
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    tests.forEach(function(test) {
        var options = test[0];
        var statusCode = test[1];
        var response = test[2];

        it([
            options.method || 'GET',
            options.path
        ].join(' '), request(options, statusCode, response));
    });
});

describe('app.use("path", middleware)', function() {
    var server;
    var app;
    var tests = [
        [{ path: '/test?file=foo' }, 200, 'Open in editor: ' + path.join(process.cwd(), 'foo')],
        [{ path: '/test?file=foo', method: 'POST' }, 200, 'Open in editor: ' + path.join(process.cwd(), 'foo')],
        [{ path: '/test?filex=foo' }, 400, '[express-open-in-editor] Parameter missed: file'],
        [{ path: '/?file=foo' }, 404, /Cannot GET \//],
        [{ path: '/xx?file=foo' }, 404, /Cannot GET \/xx/]
    ];

    beforeEach(function(done) {
        app = express();
        app.use('/test', openInEditor());
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    tests.forEach(function(test) {
        var options = test[0];
        var statusCode = test[1];
        var response = test[2];

        it([
            options.method || 'GET',
            options.path
        ].join(' '), request(options, statusCode, response));
    });
});

describe('app.get("path", middleware)', function() {
    var server;
    var app;
    var tests = [
        [{ path: '/test?file=foo' }, 200, 'Open in editor: ' + path.join(process.cwd(), 'foo')],
        [{ path: '/test?file=foo', method: 'POST' }, 404, /Cannot POST \/test/],
        [{ path: '/test?filex=foo' }, 400, '[express-open-in-editor] Parameter missed: file'],
        [{ path: '/?file=foo' }, 404, /Cannot GET \//],
        [{ path: '/xx?file=foo' }, 404, /Cannot GET \/xx/]
    ];

    beforeEach(function(done) {
        app = express();
        app.get('/test', openInEditor());
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    tests.forEach(function(test) {
        var options = test[0];
        var statusCode = test[1];
        var response = test[2];

        it([
            options.method || 'GET',
            options.path
        ].join(' '), request(options, statusCode, response));
    });
});

describe('options', function() {
    var server;
    var app;
    var tests = [
        [{ path: '/defaultCwd?file=script.js:2:12' }, 200, 'Open in editor: ' + path.join(process.cwd(), 'script.js:2:12')],
        [{ path: '/cwd?file=script.js:2:12' }, 200, 'Open in editor: ' + path.join(process.cwd(), 'some/path/script.js:2:12')]
    ];

    beforeEach(function(done) {
        app = express();
        app.get('/defaultCwd', openInEditor());
        app.get('/cwd', openInEditor({
            cwd: 'some/path'
        }));
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    tests.forEach(function(test) {
        var options = test[0];
        var statusCode = test[1];
        var response = test[2];

        it([
            options.method || 'GET',
            options.path
        ].join(' '), request(options, statusCode, response));
    });
});

describe('bad configure', function() {
    var server;
    var app;
    var originalWarn = console.warn;
    var warns = [];

    console.warn = function() {
        warns.push([].slice.call(arguments));
    };

    beforeEach(function(done) {
        app = express();
        app.use(openInEditor({ editor: 'xxx' }));
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        assert.deepEqual(warns, [
            ['[express-open-in-editor] Configure error:', 'Wrong value for `editor` option: xxx'],
            ['[express-open-in-editor] Request to open file failed, editor is not set up']
        ]);
        console.warn = originalWarn;
        server.close();
    });

    it('request should fail', request({ path: '/test?file=foo' }, 400, '[express-open-in-editor] Request to open file failed, editor is not set up'));
});

describe('bad command', function() {
    var server;
    var app;

    beforeEach(function(done) {
        app = express();
        process.env.OPEN_FILE = 'xxx';
        app.use(openInEditor());
        process.env.OPEN_FILE = OPEN_FILE;
        server = app.listen(PORT, function() {
            done();
        });
    });
    afterEach(function() {
        server.close();
    });

    it('request to server with bad command should fail', request({ path: '/test?file=foo' }, 500, new RegExp('\\[express-open-in-editor\\] ERROR: Error: Command failed: "xxx" "' + path.join(process.cwd(), 'foo:1:1') + '"')));
});


index.js
var url = require('url');
var path = require('path');
var configure = require('open-in-editor').configure;
var MESSAGE_PREFIX = '[' + require('./package.json').name + '] ';

function fail(res, code, message) {
    res.statusCode = code;
    res.end(MESSAGE_PREFIX + message);
}

module.exports = function(options) {
    var cwd = path.resolve((options && options.cwd) || process.cwd());
    var opener = configure(options || {}, function(err) {
        console.warn(MESSAGE_PREFIX + 'Configure error:', err);
    });

    return function openInEditor(req, res) {
        if (!opener) {
            var msg = 'Request to open file failed, editor is not set up';
            console.warn(MESSAGE_PREFIX + msg);
            return fail(res, 400, msg);
        }

        var filename = url.parse(req.url, true).query.file;

        if (!filename) {
            return fail(res, 400, 'Parameter missed: file');
        }

        // temporary solution
        // should take in account options.base
        filename = path.join(
            cwd,
            path
                .resolve('/', filename)
                .replace(/^[a-z]+:/i, '') // cut drive from path on Windows platform
        );

        opener.open(filename).then(
            function() {
                res.statusCode = 200;
                res.end('Open in editor: ' + filename);
            },
            function(e) {
                fail(res, 500, 'ERROR: ' + e);
            }
        );
    };
};
